File: app/src/main/java/com/tk4218/grocerylistr/Model/Ingredient.java
package com.tk4218.grocerylistr.Model;

import com.tk4218.grocerylistr.Database.JSONResult;
import com.tk4218.grocerylistr.Database.QueryBuilder;

public class Ingredient {
    private QueryBuilder mQb = new QueryBuilder();

    private int mIngredientKey;
    private String mIngredientName;
    private String mIngredientType;
    private int mShelfLife;
    private double mIngredientAmount;
    private String mIngredientUnit;

    public Ingredient(){

    }

    public Ingredient(int ingredientKey){
        JSONResult ingredient = mQb.getIngredient(ingredientKey);
        if(ingredient.getCount() > 0){
            setIngredientKey(ingredientKey);
            setIngredientName(ingredient.getString("IngredientName"));
            setIngredientType(ingredient.getString("IngredientType"));
            setShelfLife(ingredient.getInt("ShelfLife"));
        }
    }

    public Ingredient(String ingredientName){
        setIngredientName(ingredientName);
        JSONResult ingredient = mQb.getIngredientByName(ingredientName);
        if(ingredient.getCount() > 0){
            setIngredientKey(ingredient.getInt("IngredientKey"));
            setIngredientType(ingredient.getString("IngredientType"));
            setShelfLife(ingredient.getInt("ShelfLife"));
        }
    }

    public Ingredient(int recipeKey, String ingredientName){
        JSONResult ingredient = mQb.getRecipeIngredient(recipeKey, ingredientName);
        if(ingredient.getCount() > 0){
            setIngredientKey(ingredient.getInt("IngredientKey"));
            setIngredientName(ingredientName);
            setIngredientType(ingredient.getString("IngredientType"));
            setShelfLife(ingredient.getInt("ShelfLife"));
            setIngredientAmount(ingredient.getDouble("IngredientAmount"));
            setIngredientUnit(ingredient.getString("IngredientUnit"));
        }
    }

    public Ingredient(int ingredientKey, String ingredientName, String ingredientType, int shelfLife, double ingredientAmount, String ingredientUnit){
        setIngredientKey(ingredientKey);
        setIngredientName(ingredientName);
        setIngredientType(ingredientType);
        setShelfLife(shelfLife);
        setIngredientAmount(ingredientAmount);
        setIngredientUnit(ingredientUnit);
    }

    public void setIngredientKey(int ingredientKey){
        mIngredientKey = ingredientKey;
    }
    public int getIngredientKey(){
        return mIngredientKey;
    }
    public void setIngredientName(String ingredientName){
        mIngredientName = ingredientName;
    }
    public String getIngredientName(){
        return mIngredientName;
    }
    public void setIngredientType(String ingredientType){
        mIngredientType = ingredientType;
    }
    public String getIngredientType(){
        return mIngredientType;
    }
    public void setShelfLife(int shelfLife){
        mShelfLife = shelfLife;
    }
    public int getShelfLife(){
        return mShelfLife;
    }
    public void setIngredientAmount(double ingredientAmount){ mIngredientAmount = ingredientAmount; }
    public double getIngredientAmount(){
        return  mIngredientAmount;
    }
    public String getFormattedIngredientAmount() { return toFraction(mIngredientAmount, 1000);}
    public void setIngredientUnit(String ingredientUnit){ mIngredientUnit = ingredientUnit; }
    public String getIngredientUnit(){
        return mIngredientUnit;
    }

    private String toFraction(double d, int factor) {
        StringBuilder sb = new StringBuilder();
        if (d < 0) {
            sb.append('-');
            d = -d;
        }
        long l = (long) d;
        if (l != 0) sb.append(l);
        d -= l;
        double error = Math.abs(d);
        int bestDenominator = 1;
        for(int i=2;i<=factor;i++) {
            double error2 = Math.abs(d - (double) Math.round(d * i) / i);
            if (error2 < error) {
                error = error2;
                bestDenominator = i;
            }
        }
        if (bestDenominator > 1) {
            if (l != 0) sb.append(' ');
            sb.append(Math.round(d * bestDenominator)).append('/').append(bestDenominator);
        }
        return sb.toString();
    }
}

File: app/src/main/java/com/tk4218/grocerylistr/model/Ingredient.kt
package com.tk4218.grocerylistr.model

import android.content.Context
import android.util.Log
import androidx.databinding.BaseObservable
import androidx.databinding.Bindable
import com.google.firebase.database.*

import com.tk4218.grocerylistr.BR
import com.tk4218.grocerylistr.R
import kotlin.math.abs
import kotlin.math.roundToInt

class Ingredient() : BaseObservable() {
    var ingredientKey: String? = null
    private var mIngredientUnits = arrayOf<String>()
    private var mIngredientTypes = arrayOf<String>()

    constructor(context: Context): this() {
        mIngredientUnits = context.resources.getStringArray(R.array.measurements)
        mIngredientTypes = context.resources.getStringArray(R.array.ingredient_type)
    }

    @get:Bindable
    var ingredientName: String? = null
        set(ingredientName) {
            if (ingredientName != this.ingredientName) {
                field = ingredientName
                notifyPropertyChanged(BR.ingredientName)
            }
        }

    @get:Bindable
    var ingredientType: String? = null
        set(ingredientType) {
            if (ingredientType != this.ingredientType) {
                field = ingredientType
                notifyPropertyChanged(BR.ingredientType)
            }
        }

    private var shelfLifeInterval: Int = 0
    @get:Bindable
    var ingredientAmount: Double = 0.toDouble()
        set(ingredientAmount) {
            if (this.ingredientAmount != ingredientAmount) {
                field = ingredientAmount
                notifyPropertyChanged(BR.ingredientAmount)
            }
        }

    var preparation1: String? = null
        set(preparation1) {
            if (this.preparation1 != preparation1) {
                field = preparation1
            }
        }

    @get:Bindable
    var ingredientTypePosition: Int = 0
        set(position) {
            field = position
            if(mIngredientTypes.isNotEmpty())
                ingredientType = mIngredientTypes[position]
            notifyPropertyChanged(BR.ingredientTypePosition)
        }

    @get:Bindable
    var shelfLife: Int = 0
        set(shelfLife) {
            if (field != shelfLife) {
                field = shelfLife
                notifyPropertyChanged(BR.shelfLife)
            }
        }

    var shelfLifeText: String
        get() = shelfLife.toString()
        @Bindable
        set(shelfLifeText) {
            if (shelfLifeText.isNotEmpty() && Integer.parseInt(shelfLifeText) != shelfLife) {
                shelfLife = Integer.parseInt(shelfLifeText)
                notifyPropertyChanged(BR.shelfLifeText)
            }
        }

    @Exclude
    @get:Bindable
    var shelfLifeIntervalPosition: Int = 0
        set(position) {
            field = position
            when (position) {
                0 -> shelfLifeInterval = 1
                1 -> shelfLifeInterval = 7
                2 -> shelfLifeInterval = 30
            }
            notifyPropertyChanged(BR.shelfLifeIntervalPosition)
        }

    val formattedIngredientAmount: String
        get() = toFraction(ingredientAmount, 10)

    @get:Bindable
    var ingredientUnit: String = ""
        set(ingredientUnit) {
            if (ingredientUnit != field) {
                field = ingredientUnit
                notifyPropertyChanged(BR.ingredientUnit)
            }
        }

    @Exclude
    @get:Bindable
    var ingredientUnitPosition: Int = 0
        set(ingredientUnitPosition) {
            if (field != ingredientUnitPosition) {
                field = ingredientUnitPosition
                if(mIngredientUnits.isNotEmpty())
                    ingredientUnit = mIngredientUnits[ingredientUnitPosition]
                notifyPropertyChanged(BR.ingredientUnitPosition)
            }
        }

    fun save(): Boolean {
        val database = FirebaseDatabase.getInstance()
        val ingredientRef = database.getReference("ingredient")

        if (ingredientKey != null) {
            ingredientRef.child(ingredientKey!!).setValue(this)
        } else {
            val newIngredientRef = ingredientRef.push()
            ingredientKey = newIngredientRef.key
            newIngredientRef.setValue(this)
        }
        return true
    }

    private fun toFraction(number: Double, factor: Int): String {
        var d = number
        val sb = StringBuilder()
        if (d < 0) {
            sb.append('-')
            d = -d
        }
        val l = d.toLong()
        if (l != 0L) sb.append(l)
        d -= l.toDouble()
        var error = abs(d)
        var bestDenominator = 1
        for (i in 2..factor) {
            val error2 = abs(d - (d * i).roundToInt().toDouble() / i)
            if (error2 < error) {
                error = error2
                bestDenominator = i
            }
        }
        if (bestDenominator > 1) {
            if (l != 0L) sb.append(' ')
            sb.append((d * bestDenominator).roundToInt()).append('/').append(bestDenominator)
        }
        return sb.toString()
    }

    companion object {
        fun getIngredient(ingredientKey: String): Ingredient {
            val ingredient = arrayOfNulls<Ingredient>(1)
            val database = FirebaseDatabase.getInstance()
            val ingredientRef = database.getReference("recipe/$ingredientKey")
            ingredientRef.addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    ingredient[0] = dataSnapshot.getValue(Ingredient::class.java)
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    Log.e("Recipe Error", "Unable to Retrieve Recipe: $ingredientKey")
                }
            })
            return ingredient[0]!!
        }
    }
}
